### PR TITLE
=str Optimize concat for `javadsl.Source.empty`.

### PR DESCRIPTION
**Pull request by He-Pin**

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 1]
```diff
@@ -3264,7 +3264,8 @@ trait FlowOps[+Out, +Mat] {
 
   private def internalConcat[U >: Out, Mat2](that: Graph[SourceShape[U], Mat2], detached: Boolean): Repr[U] =
     that match {
-      case source if source eq Source.empty => this.asInstanceOf[Repr[U]]
+      case source: Source[U @unchecked, _] if source eq Source.empty                 => this.asInstanceOf[Repr[U]]
+      case source: javadsl.Source[U @unchecked, _] if source.asScala eq Source.empty => this.asInstanceOf[Repr[U]]
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:3268]: I was thinking we could put an internal  `isEmpty(Graph[SourceShape[_], _])` somewhere, like we have `getSingleSource` in `TraversalBuilder` (but this is unrelated to `TraversalBuilder` so somewhere else)

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:3268]: I know, but don't find a good place to

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:3268]: I just realized that we need to do something just like what getSingleSource does to be sure we don't omit mapMatVal so perhaps it isn't such a bad place after all.

**Review by johanandren — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala [hunk 2]
```diff
@@ -370,6 +370,16 @@ import akka.util.unused
         }
     }
   }
+
+  /**
+   * Test if a Graph is an empty Source.
+   * */
+  def isEmptySource(graph: Graph[SourceShape[_], _]): Boolean = graph match {
+    case source: scaladsl.Source[_, _] if source eq scaladsl.Source.empty        => true
+    case source: javadsl.Source[_, _] if source.asScala eq scaladsl.Source.empty => true
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala:379]: would it be better with `if source eq javadsl.Source.empty`?
I think that would avoid `asScala` conversion for cases when it's not empty.

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala:379]: But the default implementation of `asScala` seems just return the `delegate`, but maybe that will change one day, will change.

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala]: what I mean is that the conversion was needed for any other javadsl Source that would fall through to the `case _`

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala (hunk 2 above)
Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala]: Ohch! that's true for pf, thanks.
